src/contracts/protocol/libraries/logic/LiquidationLogic.sol

```diff
@@ -147,26 +261,87 @@ library LiquidationLogic {
+      (vars.userReserveDebt * vars.debtAssetPrice) /
+      vars.debtAssetUnit;
+
+    vars.userReserveCollateralInBaseCurrency =
+      (vars.userCollateralBalance * vars.collateralAssetPrice) /
+      vars.collateralAssetUnit;
+
+    // by default whole debt in the reserve could be liquidated
+    uint256 maxLiquidatableDebt = vars.userReserveDebt;
+    // but if debt and collateral is above or equal MIN_BASE_MAX_CLOSE_FACTOR_THRESHOLD
+    // and health factor is above CLOSE_FACTOR_HF_THRESHOLD this amount may be adjusted
+    if (
+      vars.userReserveCollateralInBaseCurrency >= MIN_BASE_MAX_CLOSE_FACTOR_THRESHOLD &&
+      vars.userReserveDebtInBaseCurrency >= MIN_BASE_MAX_CLOSE_FACTOR_THRESHOLD &&
+      vars.healthFactor > CLOSE_FACTOR_HF_THRESHOLD
+    ) {
+      uint256 totalDefaultLiquidatableDebtInBaseCurrency = vars.totalDebtInBaseCurrency.percentMul(
+        DEFAULT_LIQUIDATION_CLOSE_FACTOR
+      );
 
-    vars.userCollateralBalance = vars.collateralAToken.balanceOf(params.user);
+      // if the debt is more then DEFAULT_LIQUIDATION_CLOSE_FACTOR % of the whole,
+      // then we CAN liquidate only up to DEFAULT_LIQUIDATION_CLOSE_FACTOR %
+      if (vars.userReserveDebtInBaseCurrency > totalDefaultLiquidatableDebtInBaseCurrency) {
+        maxLiquidatableDebt =
+          (totalDefaultLiquidatableDebtInBaseCurrency * vars.debtAssetUnit) /
+          vars.debtAssetPrice;
+      }
+    }
+
+    vars.actualDebtToLiquidate = params.debtToCover > maxLiquidatableDebt
+      ? maxLiquidatableDebt
+      : params.debtToCover;
 
     (
       vars.actualCollateralToLiquidate,
       vars.actualDebtToLiquidate,
-      vars.liquidationProtocolFeeAmount
+      vars.liquidationProtocolFeeAmount,
+      vars.collateralToLiquidateInBaseCurrency
     ) = _calculateAvailableCollateralToLiquidate(
-      collateralReserve,
-      vars.debtReserveCache,
-      params.collateralAsset,
-      params.debtAsset,
+      collateralReserve.configuration,
+      vars.collateralAssetPrice,
+      vars.collateralAssetUnit,
+      vars.debtAssetPrice,
+      vars.debtAssetUnit,
       vars.actualDebtToLiquidate,
       vars.userCollateralBalance,
-      vars.liquidationBonus,
-      IPriceOracleGetter(params.priceOracle)
+      vars.liquidationBonus
     );
 
-    if (vars.userTotalDebt == vars.actualDebtToLiquidate) {
-      userConfig.setBorrowing(debtReserve.id, false);
+    // to prevent accumulation of dust on the protocol, it is enforced that you either
+    // 1. liquidate all debt
+    // 2. liquidate all collateral
+    // 3. leave more than MIN_LEFTOVER_BASE of collateral & debt
+    if (
+      vars.actualDebtToLiquidate < vars.userReserveDebt &&
+      vars.actualCollateralToLiquidate + vars.liquidationProtocolFeeAmount <
+      vars.userCollateralBalance
+    ) {
+      bool isDebtMoreThanLeftoverThreshold = ((vars.userReserveDebt - vars.actualDebtToLiquidate) *
+        vars.debtAssetPrice) /
+        vars.debtAssetUnit >=
```

Comment:
do we need this? `MIN_LEFTOVER_BASE` has same number of decimals (8)

Reply:
asking if we need to divide by asset units, as both things we are comparing are denominated in same terms (8 units)